[dnfdaemon] Package download and rpm transaction progress bars

m-blaha opened the pull request.

m-blaha committed:
dnfdaemon: Move callbacks to separate file

Also new parent class of all dbus callbacks added to handle common d-bus
specific operations.

m-blaha committed:
dnfdaemon: Clang format fixes

m-blaha committed:
dnfdaemon: Add package download callbacks

Signals are sent via dbus informing about packages download progress.

m-blaha committed:
dnfdaemon-client: Move callbacks to separate files

m-blaha committed:
dnfdaemon: Refactor callbacks structure

- unify session references across callbacks
- move some methods to parent class

m-blaha committed:
dnfdaemon: Signals on rpm transaction progress.

m-blaha committed:
dnfdaemon-client: Transaction progress bars

m-blaha committed:
dnfdaemon: Emit dbus signal after whole rpm transaction is finished

There is no rpm callback to indicate that all the work is over so create
one.

m-blaha committed:
dnfdaemon-client: Package download progress bars

dmach commented on dnfdaemon-client/callbacks.hpp:
```diff
@@ -0,0 +1,122 @@
+
+private:
+    libdnf::cli::progressbar::MultiProgressBar multi_progress_bar;
+    // map {package nevra: progressbar}
```
Could you confirm that NEVRA is sufficiently unique?
It **should**, but wouldn't be better using libsolv's `Id` if we have it?

m-blaha replied:
Right, I've switched it to the package Id. The NEVRA might not be unique for example in case of the reposync command (not implemented yet).

dmach commented on dnfdaemon-client/context.cpp:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright (C) 2020 Red Hat, Inc.
+Copyright (C) 2021 Red Hat, Inc.
```
replace with `2020-2021`

m-blaha replied:
changed.

dmach commented on dnfdaemon-client/context.hpp:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright (C) 2020 Red Hat, Inc.
+Copyright (C) 2021 Red Hat, Inc.
```
replace with `2020-2021`

dmach commented on dnfdaemon-server/transaction.hpp:
```diff
@@ -20,15 +20,17 @@ along with dnfdaemon-server.  If not, see <https://www.gnu.org/licenses/>.
 class RpmTransactionItem : public libdnf::rpm::TransactionItem {
 public:
-    enum class Actions { INSTALL, ERASE, UPGRADE, DOWNGRADE, REINSTALL };
+    enum class Actions { INSTALL, ERASE, UPGRADE, DOWNGRADE, REINSTALL, CLEANUP };
```
I just notice `ERASE` - we rather use `REMOVE` in libdnf.
Please check also include/libdnf/transaction/transaction_item_action.hpp if it has any use for you.

m-blaha replied:
Well, you are right and unifying these actions would be nice (we currently have three copies - one in libdnf/transaction and two others in microdnf and dnfdaemon). I consider this part of daemon kind of workaround - the code should be removed once we have better higher level transaction API in libdnf.

m-blaha committed:
dnfdaemon: Use libsolv's Id as package identifier in signals

Currently used NEVRA might not be sufficiently unique for example in
case of reposync.

dmach commented:
```
$ dnfdaemon-client install iftop
[0/0] Fedora - Rawhide - Developmental packages for the next Fedora release                                                                                                                         100% |   0.0   B/s |   0.0   B |  00m00s
[0/0] Fedora - Rawhide - Developmental packages for the next Fedora release                                                                                                                         100% |   1.8 MiB/s |   4.7 MiB |  00m03s
[0/0] Fedora - Modular Rawhide - Developmental packages for the next Fedora release                                                                                                                 100% |   0.0   B/s |   0.0   B |  00m00s
[0/0] Fedora - Modular Rawhide - Developmental packages for the next Fedora release                                                                                                                 100% | 555.4 KiB/s | 161.6 KiB |  00m00s
[0/0] Copr repo for dnf-nightly owned by rpmsoftwaremanagement                                                                                                                                      100% |   0.0   B/s |   0.0   B |  00m00s
[0/0] Copr repo for dnf-nightly owned by rpmsoftwaremanagement                                                                                                                                      100% |   0.0   B/s |   0.0   B |  00m01s
Action: Install
Name: iftop
Arch: x86_64
EVR: 0:1.0-0.25.pre4.fc34
Repo: rawhide
Is this ok [y/N]: y
[0/0] Copr repo for dnf-nightly owned by rpmsoftwaremanagement                                                                                                                                      100% |   0.0   B/s |   0.0   B |  00m01s
```
The last line should inform about package download and then there should be another with transaction progress.
This one refers to downloading repodata, which is (most likely) wrong.

dmach commented:
I've tested the PR, it looks good to me.
